server/etcdserver/server.go

```diff
@@ -1810,19 +1810,12 @@ func (s *EtcdServer) apply(
 // applyEntryNormal applies an EntryNormal type raftpb request to the EtcdServer
 func (s *EtcdServer) applyEntryNormal(e *raftpb.Entry) {
 	shouldApplyV3 := membership.ApplyV2storeOnly
-	applyV3Performed := false
-	defer func() {
-		// The txPostLock callback will not get called in this case,
-		// so we should set the consistent index directly.
-		if s.consistIndex != nil && !applyV3Performed && membership.ApplyBoth == shouldApplyV3 {
-			s.consistIndex.SetConsistentIndex(e.Index, e.Term)
-		}
-	}()
 	index := s.consistIndex.ConsistentIndex()
 	if e.Index > index {
 		// set the consistent index of current executing entry
 		s.consistIndex.SetConsistentApplyingIndex(e.Index, e.Term)
 		shouldApplyV3 = membership.ApplyBoth
+		defer s.consistIndex.SetConsistentIndex(e.Index, e.Term)
```

Comment:
After second though, I think that there is one condition when we would prefer to NOT advance the consistency index, i.e. when the Apply() function panics. It doesn't seems to be regression in the comparison to the current code... but still it's subtle. 

Looking at: 
https://stackoverflow.com/questions/34842322/can-i-recover-from-panic-handle-the-error-then-panic-again-and-keep-the-origin

a boolean `applied=true` might be a good way to detect such a panic.

Reply:
Thanks @ptabor raising this subtle & interesting scenario.  After second thought, I decided to revert to my original implementation. 

Firstly, let me clarify the subtle scenario, in case there is any misunderstanding.  The rough workflow (which may lead to data inconsistency) is something like below,
1. The method [s.applyV3.Apply](https://github.com/etcd-io/etcd/blob/52662ccd0665ee2a4991b4e99d679b67346608ac/server/etcdserver/server.go#L1877) panics for whatever reason; So the data is actually not persisted to db at all (still in memory before the commit); 
2. The deferred method `s.consistIndex.SetConsistentIndex(...)` gets called, accordingly the consistent_index is updated;
3. The periodical commit happens to get executed, so the `consistent_index` is persisted into db by the onPreCommitHook;
4. The etcd process crashes due the panicking. 

Eventually the data inconsistency occurs. Please let me know if you have the same understanding.

Secondly, usually the above workflow/case will not happen, but I agree we should prevent it from happening even in some extreme scenarios. But the proposal of introducing a boolean `applied=true` isn't good to me, because it complicates the implementation, and It makes the source code harder to maintain. 

When method [s.applyV3.Apply](https://github.com/etcd-io/etcd/blob/52662ccd0665ee2a4991b4e99d679b67346608ac/server/etcdserver/server.go#L1877) panics for whatever reason, then the `var ar *applyResult` (line 1813) will be nil. So I think it should can cover the Apply panicking case.

Reply:
Yes - that's the scenario. 

Indeed 'ar!=nil` is equivalent of applied=true. 

Why not ? 
```
	index := s.consistIndex.ConsistentIndex()
	if e.Index > index {
		// set the consistent index of current executing entry
		s.consistIndex.SetConsistentApplyingIndex(e.Index, e.Term)
                defer func() {
                   if ar!=nil {  //  &&`ar.err != nil` ?
                     s.consistIndex.SetConsistentIndex(e.Index, e.Term)
                   }
                }
		shouldApplyV3 = membership.ApplyBoth
	}
```

`ar.err != nil` is interesting. On one hand it must be fully deterministic, i.e. all replicas fail the same way normally and on replies... so we should bump the consistent index... 
On the other hand... I have doubts whether NO_SPACE alarm is fully deterministic. 
Maybe we should not 'bump' on `ar.err != nil` to minimize the change, and investigate it in the follow up PR.

Reply:
Ah, good suggestion. Thanks & updated. 

The only minor comment, actually there is no need to set the consistent_index again if the apply is successfuly. But it will do any harm.